brayns/engine/Scene.cpp

```diff
@@ -20,522 +21,119 @@
-    if (i == list.end())
-        return std::shared_ptr<T>{};
-    auto result = *i;
-    *i = newObj;
-    return result;
-}
-} // namespace
+#include <brayns/engine/common/DataHandler.h>
 
 namespace brayns
 {
-void Scene::commit()
+Scene::Scene()
+    : _bounds(Vector3f(0.f), Vector3f(0.f))
 {
 }
 
-Scene::Scene(AnimationParameters &animationParameters, VolumeParameters &volumeParameters)
-    : _animationParameters(animationParameters)
-    , _volumeParameters(volumeParameters)
+Scene::~Scene()
 {
+    ospRelease(_handle);
 }
 
-const Boxd &Scene::getBounds() const
+const Bounds &Scene::getBounds() const noexcept
 {
     return _bounds;
 }
 
-LightManager &Scene::getLightManager()
+void Scene::computeBounds() noexcept
 {
-    return _lightManager;
+    _bounds = _modelManager.getBounds();
 }
 
-size_t Scene::addModel(ModelDescriptorPtr modelDescriptor)
+void Scene::preRender(const ParametersManager &params)
 {
-    auto &model = modelDescriptor->getModel();
-    if (model.empty())
-        throw std::runtime_error("Empty models not supported.");
-
-    model.buildBoundingBox();
-
-    // Since models can be added concurrently we check if that is supported
-    if (supportsConcurrentSceneUpdates())
-        model.commitGeometry();
-
-    {
-        std::unique_lock<std::shared_timed_mutex> lock(_modelMutex);
-        modelDescriptor->setModelID(_modelID++);
-        _modelDescriptors.push_back(modelDescriptor);
-
-        // add default instance of this model to render something
-        if (modelDescriptor->getInstances().empty())
-            modelDescriptor->addInstance({true, true, modelDescriptor->getTransformation()});
-    }
-
-    _updateAnimationParameters();
-
-    _computeBounds();
-    markModified();
-
-    return modelDescriptor->getModelID();
+    _modelManager.preRender(params);
 }
 
-void Scene::addModel(const size_t id, ModelDescriptorPtr modelDescriptor)
+void Scene::postRender(const ParametersManager &params)
 {
-    auto &model = modelDescriptor->getModel();
-    if (model.empty())
-        throw std::runtime_error("Empty models not supported.");
-
-    model.buildBoundingBox();
-
-    // Since models can be added concurrently we check if that is supported
-    if (supportsConcurrentSceneUpdates())
-        model.commitGeometry();
-
-    {
-        if (replaceModel(id, modelDescriptor))
-            modelDescriptor->setModelID(id);
-
-        // add default instance of this model to render something
-        if (modelDescriptor->getInstances().empty())
-            modelDescriptor->addInstance({true, true, modelDescriptor->getTransformation()});
-    }
-
-    _updateAnimationParameters();
-
-    _computeBounds();
-    markModified();
+    _modelManager.postRender(params);
 }
 
-bool Scene::removeModel(const size_t id)
+bool Scene::commit()
 {
-    ModelDescriptorPtr model = nullptr;
-
-    if (supportsConcurrentSceneUpdates())
+    bool needsCommit = false;
+    if (!_handle)
```

Comment:
Just curiosity what is the purpose of these lazy initializations ?

Reply:
Changed